File: code/go/ch17/src/luago/compiler/codegen/func_info.go
package codegen

import . "luago/compiler/ast"
import . "luago/compiler/lexer"
import . "luago/vm"

var arithAndBitwiseBinops = map[int]int{
	TOKEN_OP_ADD:  OP_ADD,
	TOKEN_OP_SUB:  OP_SUB,
	TOKEN_OP_MUL:  OP_MUL,
	TOKEN_OP_MOD:  OP_MOD,
	TOKEN_OP_POW:  OP_POW,
	TOKEN_OP_DIV:  OP_DIV,
	TOKEN_OP_IDIV: OP_IDIV,
	TOKEN_OP_BAND: OP_BAND,
	TOKEN_OP_BOR:  OP_BOR,
	TOKEN_OP_BXOR: OP_BXOR,
	TOKEN_OP_SHL:  OP_SHL,
	TOKEN_OP_SHR:  OP_SHR,
}

type upvalInfo struct {
	locVarSlot int
	upvalIndex int
	index      int
}

type locVarInfo struct {
	prev     *locVarInfo
	name     string
	scopeLv  int
	slot     int
	captured bool
}

type funcInfo struct {
	parent    *funcInfo
	subFuncs  []*funcInfo
	usedRegs  int
	maxRegs   int
	scopeLv   int
	locVars   []*locVarInfo
	locNames  map[string]*locVarInfo
	upvalues  map[string]upvalInfo
	constants map[interface{}]int
	breaks    [][]int
	insts     []uint32
	numParams int
	isVararg  bool
}

func newFuncInfo(parent *funcInfo, fd *FuncDefExp) *funcInfo {
	return &funcInfo{
		parent:    parent,
		subFuncs:  []*funcInfo{},
		locVars:   make([]*locVarInfo, 0, 8),
		locNames:  map[string]*locVarInfo{},
		upvalues:  map[string]upvalInfo{},
		constants: map[interface{}]int{},
		breaks:    make([][]int, 1),
		insts:     make([]uint32, 0, 8),
		numParams: len(fd.ParList),
		isVararg:  fd.IsVararg,
	}
}

/* constants */

func (self *funcInfo) indexOfConstant(k interface{}) int {
	if idx, found := self.constants[k]; found {
		return idx
	}

	idx := len(self.constants)
	self.constants[k] = idx
	return idx
}

/* registers */

func (self *funcInfo) allocReg() int {
	self.usedRegs++
	if self.usedRegs >= 255 {
		panic("function or expression needs too many registers")
	}
	if self.usedRegs > self.maxRegs {
		self.maxRegs = self.usedRegs
	}
	return self.usedRegs - 1
}

func (self *funcInfo) freeReg() {
	if self.usedRegs <= 0 {
		panic("usedRegs <= 0 !")
	}
	self.usedRegs--
}

func (self *funcInfo) allocRegs(n int) int {
	if n <= 0 {
		panic("n <= 0 !")
	}
	for i := 0; i < n; i++ {
		self.allocReg()
	}
	return self.usedRegs - n
}

func (self *funcInfo) freeRegs(n int) {
	if n < 0 {
		panic("n < 0 !")
	}
	for i := 0; i < n; i++ {
		self.freeReg()
	}
}

/* lexical scope */

func (self *funcInfo) enterScope(breakable bool) {
	self.scopeLv++
	if breakable {
		self.breaks = append(self.breaks, []int{})
	} else {
		self.breaks = append(self.breaks, nil)
	}
}

func (self *funcInfo) exitScope() {
	pendingBreakJmps := self.breaks[len(self.breaks)-1]
	self.breaks = self.breaks[:len(self.breaks)-1]

	a := self.getJmpArgA()
	for _, pc := range pendingBreakJmps {
		sBx := self.pc() - pc
		i := (sBx+MAXARG_sBx)<<14 | a<<6 | OP_JMP
		self.insts[pc] = uint32(i)
	}

	self.scopeLv--
	for _, locVar := range self.locNames {
		if locVar.scopeLv > self.scopeLv { // out of scope
			self.removeLocVar(locVar)
		}
	}
}

func (self *funcInfo) removeLocVar(locVar *locVarInfo) {
	self.freeReg()
	if locVar.prev == nil {
		delete(self.locNames, locVar.name)
	} else if locVar.prev.scopeLv == locVar.scopeLv {
		self.removeLocVar(locVar.prev)
	} else {
		self.locNames[locVar.name] = locVar.prev
	}
}

func (self *funcInfo) addLocVar(name string) int {
	newVar := &locVarInfo{
		name:    name,
		prev:    self.locNames[name],
		scopeLv: self.scopeLv,
		slot:    self.allocReg(),
	}

	self.locVars = append(self.locVars, newVar)
	self.locNames[name] = newVar

	return newVar.slot
}

func (self *funcInfo) slotOfLocVar(name string) int {
	if locVar, found := self.locNames[name]; found {
		return locVar.slot
	}
	return -1
}

func (self *funcInfo) addBreakJmp(pc int) {
	for i := self.scopeLv; i >= 0; i-- {
		if self.breaks[i] != nil { // breakable
			self.breaks[i] = append(self.breaks[i], pc)
			return
		}
	}

	panic("<break> at line ? not inside a loop!")
}

/* upvalues */

func (self *funcInfo) indexOfUpval(name string) int {
	if upval, ok := self.upvalues[name]; ok {
		return upval.index
	}
	if self.parent != nil {
		if locVar, found := self.parent.locNames[name]; found {
			idx := len(self.upvalues)
			self.upvalues[name] = upvalInfo{locVar.slot, -1, idx}
			locVar.captured = true
			return idx
		}
		if uvIdx := self.parent.indexOfUpval(name); uvIdx >= 0 {
			idx := len(self.upvalues)
			self.upvalues[name] = upvalInfo{-1, uvIdx, idx}
			return idx
		}
	}
	return -1
}

func (self *funcInfo) closeOpenUpvals() {
	a := self.getJmpArgA()
	if a > 0 {
		self.emitJmp(a, 0)
	}
}

func (self *funcInfo) getJmpArgA() int {
	hasCapturedLocVars := false
	minSlotOfLocVars := self.maxRegs
	for _, locVar := range self.locNames {
		if locVar.scopeLv == self.scopeLv {
			for v := locVar; v != nil && v.scopeLv == self.scopeLv; v = v.prev {
				if v.captured {
					hasCapturedLocVars = true
				}
				if v.slot < minSlotOfLocVars && v.name[0] != '(' {
					minSlotOfLocVars = v.slot
				}
			}
		}
	}
	if hasCapturedLocVars {
		return minSlotOfLocVars + 1
	} else {
		return 0
	}
}

/* code */

func (self *funcInfo) pc() int {
	return len(self.insts) - 1
}

func (self *funcInfo) fixSbx(pc, sBx int) {
	i := self.insts[pc]
	i = i << 18 >> 18                  // clear sBx
	i = i | uint32(sBx+MAXARG_sBx)<<14 // reset sBx
	self.insts[pc] = i
}

func (self *funcInfo) emitABC(opcode, a, b, c int) {
	i := b<<23 | c<<14 | a<<6 | opcode
	self.insts = append(self.insts, uint32(i))
}

func (self *funcInfo) emitABx(opcode, a, bx int) {
	i := bx<<14 | a<<6 | opcode
	self.insts = append(self.insts, uint32(i))
}

func (self *funcInfo) emitAsBx(opcode, a, b int) {
	i := (b+MAXARG_sBx)<<14 | a<<6 | opcode
	self.insts = append(self.insts, uint32(i))
}

func (self *funcInfo) emitAx(opcode, ax int) {
	i := ax<<6 | opcode
	self.insts = append(self.insts, uint32(i))
}

// r[a] = r[b]
func (self *funcInfo) emitMove(a, b int) {
	self.emitABC(OP_MOVE, a, b, 0)
}

// r[a], r[a+1], ..., r[a+b] = nil
func (self *funcInfo) emitLoadNil(a, n int) {
	self.emitABC(OP_LOADNIL, a, n-1, 0)
}

// r[a] = (bool)b; if (c) pc++
func (self *funcInfo) emitLoadBool(a, b, c int) {
	self.emitABC(OP_LOADBOOL, a, b, c)
}

// r[a] = kst[bx]
func (self *funcInfo) emitLoadK(a int, k interface{}) {
	idx := self.indexOfConstant(k)
	if idx < (1 << 18) {
		self.emitABx(OP_LOADK, a, idx)
	} else {
		self.emitABx(OP_LOADKX, a, 0)
		self.emitAx(OP_EXTRAARG, idx)
	}
}

// r[a], r[a+1], ..., r[a+b-2] = vararg
func (self *funcInfo) emitVararg(a, n int) {
	self.emitABC(OP_VARARG, a, n+1, 0)
}

// r[a] = emitClosure(proto[bx])
func (self *funcInfo) emitClosure(a, bx int) {
	self.emitABx(OP_CLOSURE, a, bx)
}

// r[a] = {}
func (self *funcInfo) emitNewTable(a, nArr, nRec int) {
	self.emitABC(OP_NEWTABLE,
		a, Int2fb(nArr), Int2fb(nRec))
}

// r[a][(c-1)*FPF+i] := r[a+i], 1 <= i <= b
func (self *funcInfo) emitSetList(a, b, c int) {
	self.emitABC(OP_SETLIST, a, b, c)
}

// r[a] := r[b][rk(c)]
func (self *funcInfo) emitGetTable(a, b, c int) {
	self.emitABC(OP_GETTABLE, a, b, c)
}

// r[a][rk(b)] = rk(c)
func (self *funcInfo) emitSetTable(a, b, c int) {
	self.emitABC(OP_SETTABLE, a, b, c)
}

// r[a] = upval[b]
func (self *funcInfo) emitGetUpval(a, b int) {
	self.emitABC(OP_GETUPVAL, a, b, 0)
}

// upval[b] = r[a]
func (self *funcInfo) emitSetUpval(a, b int) {
	self.emitABC(OP_SETUPVAL, a, b, 0)
}

// r[a] = upval[b][rk(c)]
func (self *funcInfo) emitGetTabUp(a, b, c int) {
	self.emitABC(OP_GETTABUP, a, b, c)
}

// upval[a][rk(b)] = rk(c)
func (self *funcInfo) emitSetTabUp(a, b, c int) {
	self.emitABC(OP_SETTABUP, a, b, c)
}

// r[a], ..., r[a+c-2] = r[a](r[a+1], ..., r[a+b-1])
func (self *funcInfo) emitCall(a, nArgs, nRet int) {
	self.emitABC(OP_CALL, a, nArgs+1, nRet+1)
}

// return r[a](r[a+1], ... ,r[a+b-1])
func (self *funcInfo) emitTailCall(a, nArgs int) {
	self.emitABC(OP_TAILCALL, a, nArgs+1, 0)
}

// return r[a], ... ,r[a+b-2]
func (self *funcInfo) emitReturn(a, n int) {
	self.emitABC(OP_RETURN, a, n+1, 0)
}

// r[a+1] := r[b]; r[a] := r[b][rk(c)]
func (self *funcInfo) emitSelf(a, b, c int) {
	self.emitABC(OP_SELF, a, b, c)
}

// pc+=sBx; if (a) close all upvalues >= r[a - 1]
func (self *funcInfo) emitJmp(a, sBx int) int {
	self.emitAsBx(OP_JMP, a, sBx)
	return len(self.insts) - 1
}

// if not (r[a] <=> c) then pc++
func (self *funcInfo) emitTest(a, c int) {
	self.emitABC(OP_TEST, a, 0, c)
}

// if (r[b] <=> c) then r[a] := r[b] else pc++
func (self *funcInfo) emitTestSet(a, b, c int) {
	self.emitABC(OP_TESTSET, a, b, c)
}

func (self *funcInfo) emitForPrep(a, sBx int) int {
	self.emitAsBx(OP_FORPREP, a, sBx)
	return len(self.insts) - 1
}

func (self *funcInfo) emitForLoop(a, sBx int) int {
	self.emitAsBx(OP_FORLOOP, a, sBx)
	return len(self.insts) - 1
}

func (self *funcInfo) emitTForCall(a, c int) {
	self.emitABC(OP_TFORCALL, a, 0, c)
}

func (self *funcInfo) emitTForLoop(a, sBx int) {
	self.emitAsBx(OP_TFORLOOP, a, sBx)
}

// r[a] = op r[b]
func (self *funcInfo) emitUnaryOp(op, a, b int) {
	switch op {
	case TOKEN_OP_NOT:
		self.emitABC(OP_NOT, a, b, 0)
	case TOKEN_OP_BNOT:
		self.emitABC(OP_BNOT, a, b, 0)
	case TOKEN_OP_LEN:
		self.emitABC(OP_LEN, a, b, 0)
	case TOKEN_OP_UNM:
		self.emitABC(OP_UNM, a, b, 0)
	}
}

// r[a] = rk[b] op rk[c]
// arith & bitwise & relational
func (self *funcInfo) emitBinaryOp(op, a, b, c int) {
	if opcode, found := arithAndBitwiseBinops[op]; found {
		self.emitABC(opcode, a, b, c)
	} else {
		switch op {
		case TOKEN_OP_EQ:
			self.emitABC(OP_EQ, 1, b, c)
		case TOKEN_OP_NE:
			self.emitABC(OP_EQ, 0, b, c)
		case TOKEN_OP_LT:
			self.emitABC(OP_LT, 1, b, c)
		case TOKEN_OP_GT:
			self.emitABC(OP_LT, 1, c, b)
		case TOKEN_OP_LE:
			self.emitABC(OP_LE, 1, b, c)
		case TOKEN_OP_GE:
			self.emitABC(OP_LE, 1, c, b)
		}
		self.emitJmp(0, 1)
		self.emitLoadBool(a, 0, 1)
		self.emitLoadBool(a, 1, 0)
	}
}

File: code/java/ch17/src/main/java/com/github/zxh0/luago/compiler/codegen/FuncInfo.java
package com.github.zxh0.luago.compiler.codegen;

import com.github.zxh0.luago.compiler.ast.exps.FuncDefExp;
import com.github.zxh0.luago.compiler.lexer.TokenKind;
import com.github.zxh0.luago.vm.FPB;
import com.github.zxh0.luago.vm.OpCode;

import java.util.*;

import static com.github.zxh0.luago.compiler.lexer.TokenKind.*;
import static com.github.zxh0.luago.vm.Instruction.MAXARG_sBx;

class FuncInfo {

    private static final Map<TokenKind, OpCode> arithAndBitwiseBinops = new HashMap<>();
    static {
        arithAndBitwiseBinops.put(TOKEN_OP_ADD,  OpCode.ADD);
        arithAndBitwiseBinops.put(TOKEN_OP_SUB,  OpCode.SUB);
        arithAndBitwiseBinops.put(TOKEN_OP_MUL,  OpCode.MUL);
        arithAndBitwiseBinops.put(TOKEN_OP_MOD,  OpCode.MOD);
        arithAndBitwiseBinops.put(TOKEN_OP_POW,  OpCode.POW);
        arithAndBitwiseBinops.put(TOKEN_OP_DIV,  OpCode.DIV);
        arithAndBitwiseBinops.put(TOKEN_OP_IDIV, OpCode.IDIV);
        arithAndBitwiseBinops.put(TOKEN_OP_BAND, OpCode.BAND);
        arithAndBitwiseBinops.put(TOKEN_OP_BOR,  OpCode.BOR);
        arithAndBitwiseBinops.put(TOKEN_OP_BXOR, OpCode.BXOR);
        arithAndBitwiseBinops.put(TOKEN_OP_SHL,  OpCode.SHL);
        arithAndBitwiseBinops.put(TOKEN_OP_SHR,  OpCode.SHR);
    }

    static class UpvalInfo {
        int locVarSlot;
        int upvalIndex;
        int index;
    }

    static class LocVarInfo {
        LocVarInfo prev;
        String name;
        int scopeLv;
        int slot;
        int startPC;
        int endPC;
        boolean captured;
    }

    private FuncInfo parent;
    List<FuncInfo> subFuncs = new ArrayList<>();
    int usedRegs;
    int maxRegs;
    private int scopeLv;
    List<LocVarInfo> locVars = new ArrayList<>();
    private Map<String, LocVarInfo> locNames = new HashMap<>();
    Map<String, UpvalInfo> upvalues = new HashMap<>();
    Map<Object, Integer> constants = new HashMap<>();
    private List<List<Integer>> breaks = new ArrayList<>();
    List<Integer> insts = new ArrayList<>();
    List<Integer> lineNums = new ArrayList<>();
    int line;
    int lastLine;
    int numParams;
    boolean isVararg;

    FuncInfo(FuncInfo parent, FuncDefExp fd) {
        this.parent = parent;
        line = fd.getLine();
        lastLine = fd.getLastLine();
        numParams = fd.getParList() != null ? fd.getParList().size() : 0;
        isVararg = fd.isIsVararg();
        breaks.add(null);
    }

    /* constants */

    int indexOfConstant(Object k) {
        Integer idx = constants.get(k);
        if (idx != null) {
            return idx;
        }

        idx = constants.size();
        constants.put(k, idx);
        return idx;
    }

    /* registers */

    int allocReg() {
        usedRegs++;
        if (usedRegs >= 255) {
            throw new RuntimeException("function or expression needs too many registers");
        }
        if (usedRegs > maxRegs) {
            maxRegs = usedRegs;
        }
        return usedRegs - 1;
    }

    void freeReg() {
        if (usedRegs <= 0) {
            throw new RuntimeException("usedRegs <= 0 !");
        }
        usedRegs--;
    }

    int allocRegs(int n) {
        if (n <= 0) {
            throw new RuntimeException("n <= 0 !");
        }
        for (int i = 0; i < n; i++) {
            allocReg();
        }
        return usedRegs - n;
    }

    void freeRegs(int n) {
        if (n < 0) {
            throw new RuntimeException("n < 0 !");
        }
        for (int i = 0; i < n; i++) {
            freeReg();
        }
    }

    /* lexical scope */

    void enterScope(boolean breakable) {
        scopeLv++;
        if (breakable) {
            breaks.add(new ArrayList<>());
        } else {
            breaks.add(null);
        }
    }

    void exitScope(int endPC) {
        List<Integer> pendingBreakJmps = breaks.remove(breaks.size() - 1);

        if (pendingBreakJmps != null) {
            int a = getJmpArgA();
            for (int pc : pendingBreakJmps) {
                int sBx = pc() - pc;
                int i = (sBx+MAXARG_sBx)<<14 | a<<6 | OpCode.JMP.ordinal();
                insts.set(pc, i);
            }
        }

        scopeLv--;
        for (LocVarInfo locVar : new ArrayList<>(locNames.values())) {
            if (locVar.scopeLv > scopeLv) { // out of scope
                locVar.endPC = endPC;
                removeLocVar(locVar);
            }
        }
    }

    private void removeLocVar(LocVarInfo locVar) {
        freeReg();
        if (locVar.prev == null) {
            locNames.remove(locVar.name);
        } else if (locVar.prev.scopeLv == locVar.scopeLv) {
            removeLocVar(locVar.prev);
        } else {
            locNames.put(locVar.name, locVar.prev);
        }
    }

    int addLocVar(String name, int startPC) {
        LocVarInfo newVar = new LocVarInfo();
        newVar.name = name;
        newVar.prev = locNames.get(name);
        newVar.scopeLv = scopeLv;
        newVar.slot = allocReg();
        newVar.startPC = startPC;
        newVar.endPC = 0;

        locVars.add(newVar);
        locNames.put(name, newVar);

        return newVar.slot;
    }

    int slotOfLocVar(String name) {
        return locNames.containsKey(name)
                ? locNames.get(name).slot
                : -1;
    }

    void addBreakJmp(int pc) {
        for (int i = scopeLv; i >= 0; i--) {
            if (breaks.get(i) != null) { // breakable
                breaks.get(i).add(pc);
                return;
            }
        }

        throw new RuntimeException("<break> at line ? not inside a loop!");
    }

    /* upvalues */

    int indexOfUpval(String name) {
        if (upvalues.containsKey(name)) {
            return upvalues.get(name).index;
        }
        if (parent != null) {
            if (parent.locNames.containsKey(name)) {
                LocVarInfo locVar = parent.locNames.get(name);
                int idx = upvalues.size();
                UpvalInfo upval = new UpvalInfo();
                upval.locVarSlot = locVar.slot;
                upval.upvalIndex = -1;
                upval.index = idx;
                upvalues.put(name, upval);
                locVar.captured = true;
                return idx;
            }
            int uvIdx = parent.indexOfUpval(name);
            if (uvIdx >= 0) {
                int idx = upvalues.size();
                UpvalInfo upval = new UpvalInfo();
                upval.locVarSlot = -1;
                upval.upvalIndex = uvIdx;
                upval.index = idx;
                upvalues.put(name, upval);
                return idx;
            }
        }
        return -1;
    }

    void closeOpenUpvals(int line) {
        int a = getJmpArgA();
        if (a > 0) {
            emitJmp(line, a, 0);
        }
    }

    int getJmpArgA() {
        boolean hasCapturedLocVars = false;
        int minSlotOfLocVars = maxRegs;
        for (LocVarInfo locVar : locNames.values()) {
            if (locVar.scopeLv == scopeLv) {
                for (LocVarInfo v = locVar; v != null && v.scopeLv == scopeLv; v = v.prev) {
                    if (v.captured) {
                        hasCapturedLocVars = true;
                    }
                    if (v.slot < minSlotOfLocVars && v.name.charAt(0) != '(') {
                        minSlotOfLocVars = v.slot;
                    }
                }
            }
        }
        if (hasCapturedLocVars) {
            return minSlotOfLocVars + 1;
        } else {
            return 0;
        }
    }

    /* code */

    int pc() {
        return insts.size() - 1;
    }

    void fixSbx(int pc, int sBx) {
        int i = insts.get(pc);
        i = i << 18 >> 18;                  // clear sBx
        i = i | (sBx+MAXARG_sBx)<<14; // reset sBx
        insts.set(pc, i);
    }

    // todo: rename?
    void fixEndPC(String name, int delta) {
        for (int i = locVars.size() - 1; i >= 0; i--) {
            LocVarInfo locVar = locVars.get(i);
            if (locVar.name.equals(name)) {
                locVar.endPC += delta;
                return;
            }
        }
    }

    void emitABC(int line, OpCode opcode, int a, int b, int c) {
        int i = b<<23 | c<<14 | a<<6 | opcode.ordinal();
        insts.add(i);
        lineNums.add(line);
    }

    private void emitABx(int line, OpCode opcode, int a, int bx) {
        int i = bx<<14 | a<<6 | opcode.ordinal();
        insts.add(i);
        lineNums.add(line);
    }

    private void emitAsBx(int line, OpCode opcode, int a, int sBx) {
        int i = (sBx+MAXARG_sBx)<<14 | a<<6 | opcode.ordinal();
        insts.add(i);
        lineNums.add(line);
    }

    private void emitAx(int line, OpCode opcode, int ax) {
        int i = ax<<6 | opcode.ordinal();
        insts.add(i);
        lineNums.add(line);
    }

    // r[a] = r[b]
    void emitMove(int line, int a, int b) {
        emitABC(line, OpCode.MOVE, a, b, 0);
    }

    // r[a], r[a+1], ..., r[a+b] = nil
    void emitLoadNil(int line, int a, int n) {
        emitABC(line, OpCode.LOADNIL, a, n-1, 0);
    }

    // r[a] = (bool)b; if (c) pc++
    void emitLoadBool(int line, int a, int b, int c) {
        emitABC(line, OpCode.LOADBOOL, a, b, c);
    }

    // r[a] = kst[bx]
    void emitLoadK(int line, int a, Object k) {
        int idx = indexOfConstant(k);
        if (idx < (1 << 18)) {
            emitABx(line, OpCode.LOADK, a, idx);
        } else {
            emitABx(line, OpCode.LOADKX, a, 0);
            emitAx(line, OpCode.EXTRAARG, idx);
        }
    }

    // r[a], r[a+1], ..., r[a+b-2] = vararg
    void emitVararg(int line, int a, int n) {
        emitABC(line, OpCode.VARARG, a, n+1, 0);
    }

    // r[a] = emitClosure(proto[bx])
    void emitClosure(int line, int a, int bx) {
        emitABx(line, OpCode.CLOSURE, a, bx);
    }

    // r[a] = {}
    void emitNewTable(int line, int a, int nArr, int nRec) {
        emitABC(line, OpCode.NEWTABLE,
                a, FPB.int2fb(nArr), FPB.int2fb(nRec));
    }

    // r[a][(c-1)*FPF+i] = r[a+i], 1 <= i <= b
    void emitSetList(int line, int a, int b, int c) {
        emitABC(line, OpCode.SETLIST, a, b, c);
    }

    // r[a] = r[b][rk(c)]
    void emitGetTable(int line, int a, int b, int c) {
        emitABC(line, OpCode.GETTABLE, a, b, c);
    }

    // r[a][rk(b)] = rk(c)
    void emitSetTable(int line, int a, int b, int c) {
        emitABC(line, OpCode.SETTABLE, a, b, c);
    }

    // r[a] = upval[b]
    void emitGetUpval(int line, int a, int b) {
        emitABC(line, OpCode.GETUPVAL, a, b, 0);
    }

    // upval[b] = r[a]
    void emitSetUpval(int line, int a, int b) {
        emitABC(line, OpCode.SETUPVAL, a, b, 0);
    }

    // r[a] = upval[b][rk(c)]
    void emitGetTabUp(int line, int a, int b, int c) {
        emitABC(line, OpCode.GETTABUP, a, b, c);
    }

    // upval[a][rk(b)] = rk(c)
    void emitSetTabUp(int line, int a, int b, int c) {
        emitABC(line, OpCode.SETTABUP, a, b, c);
    }

    // r[a], ..., r[a+c-2] = r[a](r[a+1], ..., r[a+b-1])
    void emitCall(int line, int a, int nArgs, int nRet) {
        emitABC(line, OpCode.CALL, a, nArgs+1, nRet+1);
    }

    // return r[a](r[a+1], ... ,r[a+b-1])
    void emitTailCall(int line, int a, int nArgs) {
        emitABC(line, OpCode.TAILCALL, a, nArgs+1, 0);
    }

    // return r[a], ... ,r[a+b-2]
    void emitReturn(int line, int a, int n) {
        emitABC(line, OpCode.RETURN, a, n+1, 0);
    }

    // r[a+1] = r[b]; r[a] = r[b][rk(c)]
    void emitSelf(int line, int a, int b, int c) {
        emitABC(line, OpCode.SELF, a, b, c);
    }

    // pc+=sBx; if (a) close all upvalues >= r[a - 1]
    int emitJmp(int line, int a, int sBx) {
        emitAsBx(line, OpCode.JMP, a, sBx);
        return insts.size() - 1;
    }

    // if not (r[a] <=> c) then pc++
    void emitTest(int line, int a, int c) {
        emitABC(line, OpCode.TEST, a, 0, c);
    }

    // if (r[b] <=> c) then r[a] = r[b] else pc++
    void emitTestSet(int line, int a, int b, int c) {
        emitABC(line, OpCode.TESTSET, a, b, c);
    }

    int emitForPrep(int line, int a, int sBx) {
        emitAsBx(line, OpCode.FORPREP, a, sBx);
        return insts.size() - 1;
    }

    int emitForLoop(int line, int a, int sBx) {
        emitAsBx(line, OpCode.FORLOOP, a, sBx);
        return insts.size() - 1;
    }

    void emitTForCall(int line, int a, int c) {
        emitABC(line, OpCode.TFORCALL, a, 0, c);
    }

    void emitTForLoop(int line, int a, int sBx) {
        emitAsBx(line, OpCode.TFORLOOP, a, sBx);
    }

    // r[a] = op r[b]
    void emitUnaryOp(int line, TokenKind op, int a, int b) {
        switch (op) {
            case TOKEN_OP_NOT:  emitABC(line, OpCode.NOT,  a, b, 0); break;
            case TOKEN_OP_BNOT: emitABC(line, OpCode.BNOT, a, b, 0); break;
            case TOKEN_OP_LEN:  emitABC(line, OpCode.LEN,  a, b, 0); break;
            case TOKEN_OP_UNM:  emitABC(line, OpCode.UNM,  a, b, 0); break;
        }
    }

    // r[a] = rk[b] op rk[c]
    // arith & bitwise & relational
    void emitBinaryOp(int line, TokenKind op, int a, int b, int c) {
        if (arithAndBitwiseBinops.containsKey(op)) {
            emitABC(line, arithAndBitwiseBinops.get(op), a, b, c);
        } else {
            switch (op) {
                case TOKEN_OP_EQ: emitABC(line, OpCode.EQ, 1, b, c); break;
                case TOKEN_OP_NE: emitABC(line, OpCode.EQ, 0, b, c); break;
                case TOKEN_OP_LT: emitABC(line, OpCode.LT, 1, b, c); break;
                case TOKEN_OP_GT: emitABC(line, OpCode.LT, 1, c, b); break;
                case TOKEN_OP_LE: emitABC(line, OpCode.LE, 1, b, c); break;
                case TOKEN_OP_GE: emitABC(line, OpCode.LE, 1, c, b); break;
            }
            emitJmp(line, 0, 1);
            emitLoadBool(line, a, 0, 1);
            emitLoadBool(line, a, 1, 0);
        }
    }

}
